File: c/alg/calc/absorption_rate.c
/*
    Minha fome vem da alma;
    Meu machado é de Assis.

    SANDER, J. C. A. - 2026
*/

#include "absorption_rate.h"

void
calc_abs_rate(
    int NUM_NODES[],
    int NUM_REGS,
    int REGS[NUM_REGS],
    double AVG_FI[],
    double H[NUM_REGS],
    double SIGMA_T[],
    double SIGMA_S0[],
    double abs_rate[NUM_REGS])
{
    int i, j, num_nodes;
    int node = 0;
    double sum, sigma_a;

    for (i = 0; i < NUM_REGS; i++)
    {
        sigma_a = SIGMA_T[i] - SIGMA_S0[i];

        num_nodes = NUM_NODES[i];
        sum = 0;
        for (j = 0; j < num_nodes; j++, node++)
        {
            sum += AVG_FI[node];
        }

        abs_rate[i] = sum * sigma_a * H[i];
    }
}

File: linear-rust/src/calc/absorption_rate.rs
/*
    Minha fome vem da alma;
    Meu machado é de Assis.

    SANDER, J. C. A. - 2026
*/

fn abs_rate(
    reg_nodes: &[i32],
    h: &[f64],
    avg_fi: &[f64],
    sigma_t: &[f64],
    sigma_s0: &[f64],
) -> Vec<f64> {
    let mut avg_fi_iter = avg_fi.iter();

    reg_nodes
        .iter()
        .enumerate()
        .map(|(reg, &n_nodes)| {
            let sigma_a = sigma_t[reg] - sigma_s0[reg];

            let sum: f64 = avg_fi_iter.by_ref().take(n_nodes as usize).sum();

            sum * sigma_a * h[reg]
        })
        .collect()
}
